src/dom.ts

```diff
@@ -269,3 +268,52 @@ export function compareElementArea(a: HTMLElement, b: HTMLElement): number {
 
     return aArea - bArea
 }
+
+export const hintworthy_js_elems = []
+
+/** Adds or removes an element from the hintworthy_js_elems array of the
+ *  current tab.
+ *  This function is made available to the page's. We should be very careful
+ *  about the input it accepts and what it does with it.
+ */
+export function registerEvListenerAction(elem: EventTarget, add: boolean, event: string) {
```

Comment:
Mostly a note to self:

The element is the real risk here. It can be any arbitrary JS object and checking that it's an instanceof Element doesn't really help security (any custom class can extend Element).

What we need to do is check that elem really is in the DOM (`document.body.contains()`) and then take an X-ray copy of it to remove all expandos/redefinitions of basic functionality.

Firefox may already X-ray it and/or prevent its methods from accessing our objects, but we need to check that.

Reply:
According to the [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.exportFunction#Exporting_functions_that_take_arguments), elements are xray'd.

Just to be sure, I added `console.log(elem, (elem as any).modifiedstuff)` to `registerEvListenerAction()` and then ran this:
```js
[].map.call(document.getElementsByTagName("div"), (e) => {
    e.modifiedstuff = "modified" ;
    e.addEventListener("click", () => console.log("click"))
})
```
The console printed a lot of `<elem> undefined` and nothing else.

Reply:
Thanks, looks like real DOM elements are X-rayed, but arbitrary JS objects aren't. So I could do something like this:

```ts
class MaliciousElement extends Element {
    get nodeName() {
        // do something bad
    }
}

EventTarget.prototype.addEventListener.apply(new MaliciousElement(), ["mousedown", ()=>{}])
```

I don't think you can attach such synthetic elements to the DOM, so if we check if they're contained we're probably safe, but we will miss out on real elements that have event listeners attached before they're attached to the DOM, which could be a common case.

Reply:
Since your example didn't work, I tried my own: I left `registerEvListenerAction` exposed and did this:
```js
var MaliciousElement = function () {}
MaliciousElement.prototype = Object.create(EventTarget.prototype)
MaliciousElement.__proto__ = Object.create(Element.prototype)
MaliciousElement.prototype.addEventListener = () => console.log("Downloading malware")
MaliciousElement.prototype.getClientsRect = () => console.log("Malware installed!")
registerEvListenerAction(new MaliciousElement(), true, "click")
```
It doesn't work, xray vision removes every modification we made to the prototype.

If we somehow find a way to inject malicious objects into our content script, I think we could protect ourselves from that by lazily checking that the object is in the DOM like this:
```ts
export const hintworthy_js_elems = ((arr) => {
   return new Proxy(arr, {
      get: (arr, id) {
         if (isNaN(id - 0))
            return arr[id]
         if (id < 0 || id >= arr.length)
            return arr[id]
         if (document.body.contains(arr[id]))
            return arr[id]
         arr.splice(id, 1);
         return arr[-1]
      }
   })
})([])
```